pkg/aggregator/ckey/key.go

```diff
@@ -27,49 +23,88 @@ import (
 // Note that Agent <= 6.19.0 were using a 128 bits hash, we've switched
 // to 64 bits for better performances (map access) and because 128 bits were overkill
 // in the first place.
-// Note that we've benchmarked against xxhash64 which should be slightly faster,
-// but the Go compiler is not inlining xxhash sum methods whereas it is inlining
-// the murmur3 implementation, providing better performances overall.
 // Note that benchmarks against fnv1a did not provide better performances (no inlining).
 type ContextKey uint64
 
+/*
+#
+#// KeyGenerator generates key
+#// Not safe for concurrent usage
+#type KeyGenerator struct {
+#	buf []byte
+#}
+#
+#// NewKeyGenerator creates a new key generator
+#func NewKeyGenerator() *KeyGenerator {
+#	return &KeyGenerator{
+#		buf: make([]byte, 0, 1024),
+#	}
+#}
+#
+#// Generate returns the ContextKey hash for the given parameters.
+#// The tags array is sorted in place to avoid heap allocations.
+#func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
+#	g.buf = g.buf[:0]
+#
+#	// Sort the tags in place. For typical tag slices, we use
+#	// the in-place insertion sort to avoid heap allocations.
+#	// We default to stdlib's sort package for longer slices.
+#	// See `pkg/util/sort.go` for info on the threshold.
+#	if len(tags) < util.InsertionSortThreshold {
+#		util.InsertionSort(tags)
+#	} else {
+#		sort.Strings(tags)
+#	}
+#
+#	g.buf = append(g.buf, name...)
+#	g.buf = append(g.buf, ',')
+#	for i := 0; i < len(tags); i++ {
+#		g.buf = append(g.buf, tags[i]...)
+#		g.buf = append(g.buf, ',')
+#	}
+#	g.buf = append(g.buf, hostname...)
+#
+#	return ContextKey(murmur3.Sum64(g.buf))
+#}
+#
+*/
+
+// NewKeyGenerator creates a new key generator
+func NewKeyGenerator() *KeyGenerator {
+	return &KeyGenerator{}
+}
+
 // KeyGenerator generates key
 // Not safe for concurrent usage
 type KeyGenerator struct {
-	buf []byte
-}
+	intb uint64
 
-// NewKeyGenerator creates a new key generator
-func NewKeyGenerator() *KeyGenerator {
-	return &KeyGenerator{
-		buf: make([]byte, 0, 1024),
-	}
+	sli [512]uint64
+	idx int
 }
 
 // Generate returns the ContextKey hash for the given parameters.
 // The tags array is sorted in place to avoid heap allocations.
 func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
-	g.buf = g.buf[:0]
+	g.intb = 0xc6a4a7935bd1e995
```

Comment:
May be worth not using them anymore as member of the key generator since those are not slices, I'll look into this but anyway, `sli` has to, so even if `intb` and `idx` can be created on the stack in the function, we may have threading issue with `sli` (to be renamed).